### PR TITLE
Adds support for TLS configuration options

### PR DESCRIPTION
**Pull request by gbts**

### CODE REVIEW

File: lib/models.go [hunk 1]
```diff
@@ -68,6 +70,109 @@ func (s *Stage) UnmarshalJSON(data []byte) error {
...
+	return nil
+}
+
+type TLSCipherSuites struct {
+	Values []uint16
+}
+
+func (s *TLSCipherSuites) UnmarshalJSON(data []byte) error {
+	// From https://golang.org/pkg/crypto/tls#pkg-constants
+	suiteMap := map[string]uint16{
```
Comment (liclac) [inline lib/models.go]: Also: I'd probably use a switch on the value instead of a map (in an unexported function); that way you can get test coverage telling you which cases are properly tested.

File: lib/options_test.go [hunk 2]
```diff
@@ -76,6 +78,26 @@ func TestOptionsApply(t *testing.T) {
 		assert.True(t, opts.InsecureSkipTLSVerify.Valid)
 		assert.True(t, opts.InsecureSkipTLSVerify.Bool)
 	})
+	t.Run("TLSCipherSuites", func(t *testing.T) {
```
Comment (liclac) [inline lib/options_test.go]: Please test each individual cipher suite as well! It'd suck if we had a couple of suites that just, didn't work…

**Comment by gbts** [comment]
Hmm, this looks a bit trickier than I thought. Apparently the list of supported cipher suites in `crypto/tls` is constantly changing and there doesn't seem to be any exported way to retrieve them from the package. Any hard-coded version like the map I used or a set of switch cases will be bound to a specific version of `crypto/tls`, and the same applies to testing (which will also need some custom logic to test each cipher). I ran into this issue by using an older release that didn't have a couple of ciphers implemented.

One way to get them dynamically could be to use reflection to get all `TLS_*` constants in `crypto/tls`, but then again some of them might be disabled (and reflection doesn't sound like a great idea anyway).

Another way we could handle this would be to

**Comment by gbts** [comment]
Hmm, this looks a bit trickier than I thought. Apparently the list of supported cipher suites in `crypto/tls` is constantly changing and there doesn't seem to be any exported way to retrieve them from the package. Any hard-coded version like the `map` I used or a set of `switch` cases will be bound to a specific version of crypto/tls, and the same applies to testing (which will also need some custom logic to test each cipher). I ran into this issue by compiling with an older release that didn't have a couple of ciphers implemented.

One way to get them dynamically could be to use reflection to get all `TLS_*` constants in `crypto/tls`, but then again some of them might be disabled (and reflection doesn't sound like a great idea anyway).

Another way we could handle this would be to just export all the [IANA cipher IDs](https://www.iana.org/assignments/tls-parameters/tls-parameters.xml) to JS and just pass those IDs to `crypto/tls`, i.e. just let the user figure out what ciphers their current `crypto/tls` version supports. The docs could explain the whole situation to the user and list the ciphers that are known to be supported by the latest version of `crypto/tls`, and testing would be restricted to a few ciphers that are well supported by most versions (we don't really need to test each cipher individually anyway since they're already tested by `crypto/tls`, all we need to test for is whether our code is actually passing the list of IDs in `TLSConfig` correctly).

I'm leaning towards the second approach just because it's simpler, but I'd like to hear your input on that before making any changes.

**Comment by liclac** [comment]
Hm, here's a crazy idea that just might work:

Have all the cipher suites in a global map, the lowest Go version we support is 1.7 (which adds `context` to the standard library), so have it default to containing all the cipher suites available in that version.

Then, use build tags to compile in `init()` functions that add subsequent releases' new cipher suites to that map. One file for 1.8+, one for 1.9+, etc.

**Comment by gbts** [comment]
Sounds good, I'll give it a shot.

**Comment by gbts** [comment]
OK, I added a couple of files with build tags for 1.7 & 1.8. The 1.8 file also includes the 1.9 tag since it seems that there won't be any changes between the two. I think all of the other issues you brought up should be resolved as well, so please take a look and let me know if any other changes are necessary.

File: lib/options.go [hunk 3]
```diff
@@ -21,10 +21,97 @@
...
 )
 
+type TLSVersion struct {
+	Min int
+	Max int
+}
+
+func (v *TLSVersion) UnmarshalJSON(data []byte) error {
+	// From https://golang.org/pkg/crypto/tls/#pkg-constants
+	versionMap := map[string]int{
```
Comment (liclac) [inline lib/options.go]: This should be a global as well.

File: lib/options.go [hunk 4]
```diff
@@ -21,10 +21,97 @@
...
+			fields.Max = version
+		default:
+			return err
+		}
+	}
+
+	var minVersion int
+	var maxVersion int
+	var ok bool
+	if minVersion, ok = versionMap[fields.Min]; !ok {
```
Comment (liclac) [inline lib/options.go]: Instead of all this, just do:

```go
if min, ok := versionMap[fields.Min]; ok {
    v.Min = min
}
```

Or if you want to avoid corrupting `v` in case of invalid data, construct a new TLSVersion at the top of the function, do that, and end it with a `*v = var`.

File: lib/options.go [hunk 5]
```diff
@@ -21,10 +21,97 @@
...
+
+	v.Min = minVersion
+	v.Max = maxVersion
+
+	return nil
+}
+
+type TLSCipherSuites struct {
+	Values []uint16
+}
```
Comment (liclac) [inline lib/options.go]: You can shorten this to `type TLSCipherSuites []uint16`

**Comment by liclac** [comment]
Sorry about the slow response, there's a bit of a heatwave over the UK right now and it's making it a tad hard to focus...

**Comment by gbts** [comment]
Hi @liclac, I just pushed a commit for the changes you requested. No worries about the response time, I was AFK for most of this week so I wouldn't have pushed this earlier anyway!

**Comment by liclac** [comment]
Whoops, problem when building under Go 1.8:

```
lib/tlsconfig_go1_8.go:32: SupportedTLSVersions redeclared in this block
	previous declaration at lib/tlsconfig_go1_7.go:34
lib/tlsconfig_go1_8.go:52: SupportedTLSCipherSuites redeclared in this block
	previous declaration at lib/tlsconfig_go1_7.go:54
```

**Comment by gbts** [comment]
Added #265 with a fix for that
